src/Interpreters/Aggregator.cpp

```diff
@@ -1498,13 +1501,25 @@ bool Aggregator::executeOnBlock(Columns columns,
     /** Flush data to disk if too much RAM is consumed.
       * Data can only be flushed to disk if a two-level aggregation structure is used.
       */
-    if (params.max_bytes_before_external_group_by
-        && result.isTwoLevel()
-        && current_memory_usage > static_cast<Int64>(params.max_bytes_before_external_group_by)
-        && worth_convert_to_two_level)
+    if (worth_convert_to_two_level && result.isTwoLevel())
     {
-        size_t size = current_memory_usage + params.min_free_disk_space;
-        writeToTemporaryFile(result, size);
+        bool write_temporary_file = false;
+        if (params.max_bytes_before_external_group_by && current_memory_usage > static_cast<Int64>(params.max_bytes_before_external_group_by))
+            write_temporary_file = true;
+        if (params.max_bytes_ratio_before_external_group_by > 0. && current_memory_usage > memory_limit * params.max_bytes_ratio_before_external_group_by)
```

Comment:
It does make sense, however I think that per-query ratio can also be useful, since this will allow to install it to i.e. `0.5` and this will allow to avoid `MEMORY_LIMIT_EXCEEDED` errors, while it does not make a lot of sense to set ratio for memory available in total to `0.5`

So I have two options then:
- add server setting, i.e. `max_bytes_ratio_before_external_group_by_for_server`/`max_bytes_ratio_before_external_sort_for_server`
- or make this limits to trigger on the total memory consumption as well, i.e. if it was set to `0.5` then it will start using disk once either query memory limits reached 50% or the server limit

Thoughts?

Reply:
Added two new server settings to control this behavior server-wide.

Reply:
I was actually thinking about finding the difference between user memory tracker hard limit and user memory usage.

Seems like it will exactly memory_limit * params.max_bytes_ratio_before_external_group_by if there is no queries running right now and automatically cover case of multiple queries.